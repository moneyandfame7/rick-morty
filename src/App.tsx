import React from "react";
import { useLocation, useOutlet } from "react-router";
import { createBrowserRouter } from "react-router-dom";
import { routesConfig } from "./routes/routesConfig";
import { getDesignTokens } from "./utils/theme";
import { CSSTransition, SwitchTransition } from "react-transition-group";
import { ThemeProvider, createTheme } from "@mui/material/styles";
import { Container, CssBaseline, useMediaQuery } from "@mui/material";
import MyBreadcrumbs from "./components/Breadcrumbs";
import Header from "./layout/Header";

import "./App.scss";

export const ColorModeContext = React.createContext({ toggleColorMode: () => {} });

export const router = createBrowserRouter([
  {
    path: "/",
    element: <App />,
    children: routesConfig.map(route => ({
      index: route.path === "/",
      path: route.path === "/" ? undefined : route.path,
      element: route.element,
    })),
  },
]);

function App() {
  const location = useLocation();
  const currentOutlet = useOutlet();
  const { nodeRef } = routesConfig.find(route => route.path === location.pathname) ?? {};

  const prefersDarkMode = useMediaQuery("(prefers-color-scheme: dark)");
  const [mode, setMode] = React.useState<"light" | "dark">(prefersDarkMode ? "dark" : "light");
  const theme = createTheme(getDesignTokens(mode));
  const colorMode = React.useMemo(
    () => ({
      toggleColorMode: () => {
        setMode(prevMode => (prevMode === "light" ? "dark" : "light"));
      },
    }),
    []
  );

  return (
    <ColorModeContext.Provider value={colorMode}>
      <ThemeProvider theme={theme}>
        <CssBaseline />
        <Header />
        <MyBreadcrumbs />
        <Container sx={{ width: "100%", padding: 0 }}>
          <SwitchTransition>
            <CSSTransition key={location.pathname} nodeRef={nodeRef} timeout={300} classNames='page' unmountOnExit>
              {() => (
                <div
                  ref={nodeRef}
                  style={{
                    display: "flex",
                    flexDirection: "column",
                    alignItems: "center",
                    gap: "20px",
                    margin: "20px 0 50px",
                    padding: 0,
                  }}
                  className='page'
                >
                  {currentOutlet}
                </div>
              )}
            </CSSTransition>
          </SwitchTransition>
        </Container>
      </ThemeProvider>
    </ColorModeContext.Provider>
  );
}

export default App;
