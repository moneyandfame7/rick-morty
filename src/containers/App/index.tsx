import React from "react"
import { getApiResource } from "../../utils/fetch"

const App: React.FC = () => {
  return <h1>hello world</h1>
}

export default App
