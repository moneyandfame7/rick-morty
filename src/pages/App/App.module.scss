.header {
  color: blue;
}
.text {
  font-weight: 500;
  text-transform: uppercase;
}
