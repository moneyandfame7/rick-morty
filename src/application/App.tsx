import React from 'react'
import { getDesignTokens } from '../utils/theme'
import { createTheme, ThemeProvider } from '@mui/material/styles'
import { CssBaseline, useMediaQuery } from '@mui/material'
import Header from '../shared/layout/Header/default/Header'
import '../App.scss'
import { Route, Routes } from 'react-router-dom'
import { protectedRoutes, publicRoutes } from '../routes/routesConfig'
import { ProtectedRoute } from '../components/ProtectedRoute'

export const ColorModeContext = React.createContext({
  toggleColorMode: () => {}
})
// TODO: протестировать тему, кликая на кнопку смены темы по несколько раз
// export const router = createBrowserRouter([
//   {
//     path: "/",
//     element: <App />,
//     children: routesConfig.map(route => ({
//       index: route.path === "/",
//       path: route.path === "/" ? undefined : route.path,
//       element: route.element,
//     })),
//   },
// ]);

function App() {
  const prefersDarkMode = useMediaQuery('(prefers-color-scheme: dark)')
  // const location = useLocation();
  // const currentOutlet = useOutlet();
  // const { nodeRef } = routesConfig.find(route => route.path === location.pathname) ?? {};

  const [mode, setMode] = React.useState<'light' | 'dark'>(prefersDarkMode ? 'dark' : 'light')
  const theme = createTheme(getDesignTokens(mode))
  const colorMode = React.useMemo(
    () => ({
      toggleColorMode: () => {
        setMode(prevMode => (prevMode === 'light' ? 'dark' : 'light'))
      }
    }),
    []
  )
  return (
    <ColorModeContext.Provider value={colorMode}>
      <ThemeProvider theme={theme}>
        <CssBaseline />
        <Header />
        {/*{!(pathname === "/signup" || pathname === "/login" || pathname === "/welcome") ? <MyBreadcrumbs /> : null}*/}
        {/*     <SwitchTransition>
          <CSSTransition key={location.pathname} nodeRef={nodeRef} timeout={300} classNames='page' unmountOnExit>
            {() => (
              <div
                ref={nodeRef}
                style={{
                  display: "flex",
                  flexDirection: "column",
                  alignItems: "center",
                  gap: "20px",
                  margin: "0 50px",
                  padding: 0,
                }}
                className='page'
              >*/}
        <Routes>
          {protectedRoutes.map(route => (
            <Route
              index={route.index}
              path={route.path}
              key={route.id}
              element={<ProtectedRoute>{route.element}</ProtectedRoute>}
            />
          ))}
          {publicRoutes.map(route => (
            <Route path={route.path} key={route.id} element={route.element} />
          ))}
        </Routes>
        {/*{currentOutlet}*/}
        {/*          </div>
            )}
          </CSSTransition>
        </SwitchTransition>*/}
      </ThemeProvider>
    </ColorModeContext.Provider>
  )
}

export default App
