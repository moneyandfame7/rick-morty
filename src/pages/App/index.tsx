import React from "react";
import Header from "../../layouts/Header";
import { createBrowserRouter } from "react-router-dom";
import { routesConfig } from "../../routes/routesConfig";
import Wrapper from "../../layouts/Wrapper";
import { useLocation, useOutlet } from "react-router";
import { CSSTransition, SwitchTransition } from "react-transition-group";
import "./App.scss";

export const router = createBrowserRouter([
  {
    path: "/",
    element: <App />,
    children: routesConfig.map(route => ({
      index: route.path === "/",
      path: route.path === "/" ? undefined : route.path,
      element: route.element,
    })),
  },
]);
function App() {
  const location = useLocation();
  const currentOutlet = useOutlet();
  const { nodeRef } = routesConfig.find(route => route.path === location.pathname) ?? {};
  return (
    <>
      <Header />
      <Wrapper>
        <SwitchTransition>
          <CSSTransition key={location.pathname} nodeRef={nodeRef} timeout={300} classNames='page' unmountOnExit>
            {() => (
              <div ref={nodeRef} className='page'>
                {currentOutlet}
              </div>
            )}
          </CSSTransition>
        </SwitchTransition>
      </Wrapper>
    </>
  );
}

export default App;
