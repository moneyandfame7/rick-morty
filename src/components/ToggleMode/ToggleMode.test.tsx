import { ToggleMode } from "./index";
import { render, screen } from "@testing-library/react";
import React from "react";
import userEvent from "@testing-library/user-event";
import { ColorModeContext } from "../../App";

describe("ToggleMode", () => {
  let component: any;
  const mockToggleColorMode = jest.fn();

  it("should render properly", () => {
    givenComponent();

    thenItRendersProperly();
  });

  it("should change theme when click on toggle", async () => {
    givenComponent();

    const btn = screen.getByTestId("theme-button");
    userEvent.click(btn);

    const darkIcon = await screen.findByTestId("DarkModeIcon");
    userEvent.click(darkIcon);

    expect(mockToggleColorMode).toHaveBeenCalledTimes(2);

    thenItRendersProperly();
  });

  function givenComponent(props?: any) {
    component = render(
      <ColorModeContext.Provider value={{ toggleColorMode: mockToggleColorMode }}>
        <ToggleMode />
      </ColorModeContext.Provider>
    );
  }

  function thenItRendersProperly() {
    expect(component).toMatchSnapshot();
  }
});
