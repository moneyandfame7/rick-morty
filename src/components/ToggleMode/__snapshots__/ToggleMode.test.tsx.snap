// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleMode should change theme when click on toggle 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-whjmaz-MuiButtonBase-root-MuiIconButton-root"
        data-testid="theme-button"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="DarkModeIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 3c-4.97 0-9 4.03-9 9s4.03 9 9 9 9-4.03 9-9c0-.46-.04-.92-.1-1.36-.98 1.37-2.58 2.26-4.4 2.26-2.98 0-5.4-2.42-5.4-5.4 0-1.81.89-3.42 2.26-4.4-.44-.06-.9-.1-1.36-.1z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        >
          <span
            class="css-y4cjyz-MuiTouchRipple-ripple MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
            style="width: 1px; height: 1px; top: -0.5px; left: -0.5px;"
          >
            <span
              class="MuiTouchRipple-child MuiTouchRipple-childLeaving"
            />
          </span>
          <span
            class="css-y4cjyz-MuiTouchRipple-ripple MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
            style="width: 1px; height: 1px; top: -0.5px; left: -0.5px;"
          >
            <span
              class="MuiTouchRipple-child MuiTouchRipple-childLeaving"
            />
          </span>
        </span>
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-whjmaz-MuiButtonBase-root-MuiIconButton-root"
      data-testid="theme-button"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="DarkModeIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 3c-4.97 0-9 4.03-9 9s4.03 9 9 9 9-4.03 9-9c0-.46-.04-.92-.1-1.36-.98 1.37-2.58 2.26-4.4 2.26-2.98 0-5.4-2.42-5.4-5.4 0-1.81.89-3.42 2.26-4.4-.44-.06-.9-.1-1.36-.1z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      >
        <span
          class="css-y4cjyz-MuiTouchRipple-ripple MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
          style="width: 1px; height: 1px; top: -0.5px; left: -0.5px;"
        >
          <span
            class="MuiTouchRipple-child MuiTouchRipple-childLeaving"
          />
        </span>
        <span
          class="css-y4cjyz-MuiTouchRipple-ripple MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
          style="width: 1px; height: 1px; top: -0.5px; left: -0.5px;"
        >
          <span
            class="MuiTouchRipple-child MuiTouchRipple-childLeaving"
          />
        </span>
      </span>
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`ToggleMode should render properly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-whjmaz-MuiButtonBase-root-MuiIconButton-root"
        data-testid="theme-button"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="DarkModeIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 3c-4.97 0-9 4.03-9 9s4.03 9 9 9 9-4.03 9-9c0-.46-.04-.92-.1-1.36-.98 1.37-2.58 2.26-4.4 2.26-2.98 0-5.4-2.42-5.4-5.4 0-1.81.89-3.42 2.26-4.4-.44-.06-.9-.1-1.36-.1z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-whjmaz-MuiButtonBase-root-MuiIconButton-root"
      data-testid="theme-button"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="DarkModeIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 3c-4.97 0-9 4.03-9 9s4.03 9 9 9 9-4.03 9-9c0-.46-.04-.92-.1-1.36-.98 1.37-2.58 2.26-4.4 2.26-2.98 0-5.4-2.42-5.4-5.4 0-1.81.89-3.42 2.26-4.4-.44-.06-.9-.1-1.36-.1z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
